migrator/src/main/scala/akka/persistence/jdbc/migrator/JournalMigrator.scala

```diff
@@ -0,0 +1,177 @@
+
+  import profile.api._
+
+  // get the various configurations
+  private val journalConfig = new JournalConfig(system.settings.config.getConfig("jdbc-journal"))
+  private val readJournalConfig = new ReadJournalConfig(system.settings.config.getConfig("jdbc-read-journal"))
+
+  // the journal database
+  private val journaldb =
+    SlickExtension(system).database(system.settings.config.getConfig("jdbc-read-journal")).database
+
+  // get an instance of the new journal queries
+  private val newJournalQueries: JournalQueries =
+    new JournalQueries(profile, journalConfig.eventJournalTableConfiguration, journalConfig.eventTagTableConfiguration)
+
+  // let us get the journal reader
+  private val serialization = SerializationExtension(system)
+  private val legacyJournalQueries = new ReadJournalQueries(profile, readJournalConfig)
+  private val serializer = new ByteArrayJournalSerializer(serialization, readJournalConfig.pluginConfig.tagSeparator)
+
+  private val bufferSize: Int = journalConfig.daoConfig.bufferSize
+  private val parallelism: Int = journalConfig.daoConfig.parallelism
+
+  // get the journal ordering based upon the schema type used
+  private val journalOrdering: JournalOrdering = databaseType match {
+    case Postgres  => Postgres(journalConfig, newJournalQueries, journaldb)
+    case H2        => MySQL(journalConfig, newJournalQueries, journaldb)
+    case MySQL     => SqlServer(journalConfig, newJournalQueries, journaldb)
+    case Oracle    => Oracle(journalConfig, newJournalQueries, journaldb)
+    case SqlServer => H2(journalConfig, newJournalQueries, journaldb)
+  }
+
+  /**
+   * write all legacy events into the new journal tables applying the proper serialization
+   */
+  def migrate(): Future[Unit] = {
+    val query: DBIOAction[Seq[legacy.JournalRow], Streaming[legacy.JournalRow], Effect.Read with Effect.Transactional] =
+      legacyJournalQueries.JournalTable.result
+        .withStatementParameters(
+          rsType = ResultSetType.ForwardOnly,
+          rsConcurrency = ResultSetConcurrency.ReadOnly,
+          fetchSize = bufferSize)
+        .transactionally
+
+    val eventualDone: Future[Done] = Source
+      .fromPublisher(journaldb.stream(query))
+      .via(serializer.deserializeFlow)
+      .map {
+        case Success((repr, tags, ordering)) if tags.nonEmpty =>
+          repr.withPayload(Tagged(repr, tags)) -> ordering // only wrap in `Tagged` if needed
+        case Success((repr, _, ordering)) => repr -> ordering // noops map
+        case Failure(exception)           => throw exception // blow-up on failure
+      }
+      // since the write is done one at a time we can at least enhance throughput by
+      // spawning more actors to execute the db write
+      // FIXME we can figure out a way to do some batch inserts because the ordering of insertion does not matter any longer
+      .mapAsync(parallelism) { case (repr, ordering) =>
```

Comment:
although we are preserving the ordering, I would just insert in the order we read and then set this to `mapAsync(1)`. 

The natural order for many DBs, if I'm not mistaking, is the order that bytes are saved on disk. If we use `mapAsync(8)` we will interleave then on disk. We don't need to do it.

For the record, I create a journal with 500k events, each events with 200 bytes of payload. The migration took around 20 minutes with `mapAsync(1)`.

Reply:
Ok we can also group them and do some batch insertions.